lib/HLSL/DxilSimpleGVNHoist.cpp

```diff
@@ -585,3 +593,197 @@ FunctionPass *llvm::createDxilSimpleGVNHoistPass() {
+//   (Same with 'S1' and 'P1').
+//   This means if 'Br'->'S0' is taken, then 'End' must be reached via 'P0'; if
+//   'End' is reached via 'P0', 'Br'->'S0' must have been taken (Same with 'S1'
+//   and 'P1').
+// - Using ValueTable, compare if every pair of incoming values from P0 and P1
+//   is identical for any PHIs in 'End'
+// - Make sure there is no side effect or loop between 'Br' and 'End'
+// - If all above checks succeed, replace 'Br' with with an unconditional branch
+//   to S0
+// 
+// The current state of the pass is pretty limited. If incoming values from P0
+// and P1 are dependent on any PHIs defined between Br and End, then the pass
+// will fail to simplify the branch. If there are any side effects within the
+// region, the pass will also fail. It's possible to handling these cases, but
+// require proving the two sides of the branch have equivalent control flow,
+// which is non-trivial, and will be left to a later date.
+//
+namespace {
+
+class DxilSimpleGVNEliminateRegion : public FunctionPass {
+
+  bool CheckRegionForSideEffectsOrLoops(BasicBlock *Begin, BasicBlock *End /*Non inclusive*/);
+
+  std::unordered_map<BasicBlock *, bool> BlockHasSideEffects;
+  bool MayHaveSideEffects(BasicBlock *BB) {
+    auto It = BlockHasSideEffects.find(BB);
+    if (It != BlockHasSideEffects.end())
+      return It->second;
+
+    for (Instruction &I : *BB) {
+      if (I.mayHaveSideEffects() && !hlsl::IsNop(&I)) {
+        BlockHasSideEffects[BB] = true;
+        return true;
+      }
+    }
+
+    BlockHasSideEffects[BB] = false;
+    return false;
+  }
+  bool ProcessBB(BasicBlock &BB, ValueTable &VT, DominatorTree *DT, PostDominatorTree *PDT);
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilSimpleGVNEliminateRegion() : FunctionPass(ID) {}
+
+  StringRef getPassName() const override {
+    return "DXIL simple GVN eliminate region";
+  }
+
+  bool runOnFunction(Function &F) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<PostDominatorTree>();
+    AU.addRequired<DominatorTreeWrapperPass>();
+  }
+};
+
+char DxilSimpleGVNEliminateRegion::ID = 0;
+
+bool DxilSimpleGVNEliminateRegion::CheckRegionForSideEffectsOrLoops(BasicBlock *Begin, BasicBlock *End /*Non inclusive*/) {
+  SmallVector<BasicBlock *, 10> Worklist;
+  Worklist.push_back(Begin);
+  SmallPtrSet<BasicBlock *, 10> Seen;
+
+  while (Worklist.size()) {
+    BasicBlock *BB = Worklist.pop_back_val();
+    // Stop before reaching End
+    if (BB == End)
+      continue;
+
+    if (MayHaveSideEffects(BB))
+      return false;
+
+    Seen.insert(BB);
+    for (BasicBlock *Succ : successors(BB)) {
+      // Goes back into the region. Give up.
+      if (Seen.count(Succ))
+        return false;
+      Worklist.push_back(Succ);
+    }
+  }
+  return true;
+}
+
+bool DxilSimpleGVNEliminateRegion::ProcessBB(BasicBlock &BB, ValueTable &VT, DominatorTree *DT, PostDominatorTree *PDT) {
```

Comment:
This seems to still have the problem of repeatedly walking the CFG for each region. If we have a graph that contains a bunch of nested diamond graphs then it seems like would could run into pathological compile times.


```
                                A
                             /     \
                            C        G
                          /   \    /   \
                         D     E  H     I
                         \     /   \   /
                            F        J
                              \     /
                                 K
```

Reply:
I did some stress test, using a diamond graph 13 levels deep (8192 branches). The runtime is completely dominated by the linear initial attempt to identify diamond graphs. We'll have to push the program size pretty far to have that N^2 walk be a problem.